---
title: "john18nonone"
output: html_document
---



Here I analyse john's data using only subgroup 4.
```{r}
library('knitr')

knitr::opts_chunk$set(cache=TRUE)
```



```{r,echo=F,eval=T}
x <- read.table("~/Dropbox/john/s1table.txt", header = TRUE)
y=x[,-c(1,2)]
anno <- read.table("~/Dropbox/john/annotation.txt", header = TRUE, stringsAsFactors = FALSE)
head(anno)

df_ordered <- y[, order(anno$time, anno$bact, anno$ind)]
anno_ordered <- anno[order(anno$time, anno$bact, anno$ind), ]

eighteenguys=which(anno_ordered$time==18&anno_ordered$bact!="none")

anno_ordered=anno_ordered[eighteenguys,]
df_ordered=df_ordered[,eighteenguys]
```
Now, let's quantile normalize these counts:

```{r,echo=T,eval=T}


voomdat=apply(df_ordered,2,function(x){qqnorm(x,plot.it = F)$x})

##show this data is quantil normalised
#head(apply(voomdat,2,summary))
```
And transpose it so that the individuals are in the rows and the genes are in the columns:

```{r,echo=T,eval=T}
samplesbygenes=t(voomdat)
dim(samplesbygenes)##rows are the samples, columns are the genes
````

Then, subtract the column mean (e.g., the mean expression for the gene) from each column.
```{r}
rep.row<-function(x,n){
  matrix(rep(x,each=n),nrow=n)
}


mean_voom_features <- apply(samplesbygenes, 2, mean);###subtract gene(column) means
voom_class_adj <- samplesbygenes - rep.row(mean_voom_features, dim(samplesbygenes)[1])##subtract the mean expression across genes for the gene
dim(voom_class_adj)
```

Check to make sure that this has occured correctly, e.g., the *voom_class_adjusted matrix* represents any individuals quantile normalized gene expression minus the mean gene expression for that gene across individuals. Let's check for individual 10 and gene 16:

```{r}
n=10##individual 10
j=16
###check to make sure they match
samplesbygenes[n,j]-mean(samplesbygenes[,j])
voom_class_adj[n,j]
```

Now, create a model matrix that combines the time and bacteria into a factor:

```{r}
ID=with(anno_ordered,interaction(time,bact),drop=T)
m=model.matrix(~as.factor(ID)-1)

##check sample id matches strain and time
rownames(samplesbygenes)[10]
which(m[10,]==1)

````


Now fit the model for all genes to each level of the 27 subgroups iteratively. Essentially, performing univariate regression for all genes on each time*bacterial combination:

```{r,echo=TRUE,eval=F}
beta=matrix(NA,nrow = dim(voom_class_adj)[2],ncol=dim(m)[2])##make this the J genes by k subgroup betas
se=matrix(NA,nrow = dim(voom_class_adj)[2],ncol=dim(m)[2])
t=matrix(NA,nrow = dim(voom_class_adj)[2],ncol=dim(m)[2])
colnames(se)=colnames(beta)=colnames(t)=as.matrix(levels(ID))
for(k in 1:ncol(m)){
  fit=lm((voom_class_adj)~(m[,k]-1))
  a=matrix(unlist(coef(summary(fit))),byrow = T,nrow=ncol(voom_class_adj))
  beta[,k]=a[,1]
  se[,k]=a[,2]
  t[,k]=a[,3]
}


write.table(beta,"betafit18nonefree.txt",col.names = T)
write.table(se,"sefit18nonefree.txt",col.names = T)
write.table(t,"tfit18nonefree.txt",col.names = T)

##check to make sure

lm(voom_class_adj[,37]~m[,8]-1)##model the expression of gene 37 on factor 9
t[37,9]
##show that beta is equal to the mean for those individuals

mean(voom_class_adj[which(m[,8]==1),37])##mean gene expression for time point 18 individuals at gene 37
beta[37,8]
```

Now fit with univariate ash to choose maxes:

```{r,echo=TRUE,eval=F}
t=read.table("tfit18nonefree.txt")
b=read.table("betafit18nonefree.txt")
se=read.table("sefit18nonefree.txt")

s.j=se/se

lf.ash=matrix(NA,ncol=ncol(t),nrow=nrow(t))

for(i in 1:ncol(t)){
  lf.ash[,i]=ash(betahat = t[,i],sebetahat = s.j[,i],mixcompdist = "normal")$lfsr
}


thresh=0.05

index=which(rowSums(lf.ash<thresh)>0)

write.table(t[index,],"maxttime18nonefree.txt",col.names = F ,row.names = F)
write.table(lf.ash,"lfash18nonefree.txt")

system('/Users/sarahurbut/miniconda3/bin/sfa -gen maxttime18nonefree.txt -g 8207 -k 5 -n 8 i -o johntimepoint18nonefree')

A="johntimepoint18nonefree"

factor.mat=as.matrix(read.table("johntimepoint18nonefree_F.out"))
lambda.mat=as.matrix(read.table("johntimepoint18nonefree_lambda.out"))


library('mash')
cov=compute.covmat(b.gp.hat = t,sebetahat = s.j,Q =5, t.stat=t[index,],lambda.mat=lambda.mat,P=3,A=A, factor.mat = factor.mat,bma = TRUE,zero=TRUE,power=1)$covmat

omega=compute.covmat(b.gp.hat = t,sebetahat = s.j,Q =5, t.stat=t[index,],lambda.mat=lambda.mat,P=3,A=A, factor.mat = factor.mat,bma = TRUE,zero=TRUE,power=1)$omega

compute.hm.train.log.lik.pen(train.b = t,se.train = s.j,covmat = cov,A=A,pen=1)


pis=readRDS(paste0("pis",A,".rds"))$pihat
b.test=t
se.test=s.j
weightedquants=lapply(seq(1:nrow(b.test)),function(j){total.quant.per.snp(j,cov,b.gp.hat=b.test,se.gp.hat = se.test,pis,A=A,checkpoint = FALSE)})
```

Now let's plot the most interesting covariance matrix:

We note that the non-pathogenic (vaccine strains) are correlated, as are the active strains (RV and GC) and the facultative intracellular pathogens (Yers and Salm) stand out.


```{r covmats,echo=TRUE,eval=T}
A="johntimepoint18nonefree"
library('mash')
pis=readRDS(paste0("pis",A,".rds"))$pihat
t=read.table("tfit18nonefree.txt")
lf.ash=read.table('lfash18nonefree.txt')
cov=readRDS("covmatjohntimepoint18nonefree.rds")
library('gplots')
library('colorRamps')

pi.mat=matrix(pis[-length(pis)],byrow = T,ncol=10+ncol(m))
barplot(colSums(pi.mat),names=c(seq(1:9),as.matrix(levels(ID))[,1],"All"),las=2)
mostimportant=order(colSums(pi.mat),decreasing = T)[1]
for(k in mostimportant){
  x=cov[[k]]/max(diag(cov[[k]]))
colnames(x)=rownames(x)=as.matrix(levels(ID))[,1]
  heatmap.2(x,revC = T,col = blue2green(256),dendrogram="none",density="none",trace="none")
v=svd(cov[[k]])$v
barplot(v[,1]/max(v[,1]),names=as.matrix(levels(ID))[,1],las=2)
barplot(v[,2]/max(v[,2]),names=as.matrix(levels(ID))[,1],las=2)
barplot(v[,3]/max(v[,3]),names=as.matrix(levels(ID))[,1],las=2)
barplot(v[,4]/max(v[,3]),names=as.matrix(levels(ID))[,1],las=2)
barplot(v[,5]/max(v[,3]),names=as.matrix(levels(ID))[,1],las=2)
}
lfsr=read.table("johntimepoint18nonefreelfsr.txt")[,-1]
postt=read.table("johntimepoint18nonefreeposterior.means.txt")[,-1]
mean(lfsr<0.05)
mean(lf.ash<0.05)
##and this percent of values are shrunk###
mean(abs(postt)<abs(t))
```

CHecking

```{r,eval=F}
k=which.max(pis)
b.test=t
se.test=t/t
all.arrays=post.array.per.snp(j=j,covmat = cov,b.gp.hat = b.test,se.gp.hat = se.test)

b.mle=as.vector(t(b.test[j,]))##turn i into a R x 1 vector
V.gp.hat=diag(se.test[j,])^2
V.gp.hat.inv <- solve(V.gp.hat)


U.gp1kl <- (post.b.gpkl.cov(V.gp.hat.inv, cov[[k]]))
mu.gp1kl <- as.array(post.b.gpkl.mean(b.mle, V.gp.hat.inv, U.gp1kl))
source("~/matrix_ash/R/sarah's_mixem.R")
bovy.mean=post.b.jk.ed.mean(b.mle = b.mle,tinv = solve(V.gp.hat+cov[[k]]),U.k = cov[[k]])
plot(bovy.mean,mu.gp1kl)
bovy.cov=post.b.jk.ed.cov(tinv =solve(V.gp.hat+cov[[k]]),U.k = cov[[k]] )
plot(diag(bovy.cov),diag(U.gp1kl))
#(all.arrays$post.means[k,]) To show that our computation is correct
plot(all.arrays$post.means[k,],mu.gp1kl)
plot(all.arrays$post.covs[k,],diag(U.gp1kl))

##and here we show that our weighting is correct:

##Now, check to make sure weighting is correct

mash.means=postt
lfsr.mash=lfsr

log.lik.snp=log.lik.func(b.mle,V.gp.hat,cov)
log.lik.minus.max=log.lik.snp-max(log.lik.snp)

exp.vec=exp(log.lik.minus.max)
post.weights=t(exp.vec*pis/sum(exp.vec*pis))
plot(as.matrix(mash.means[j,]),post.weights%*%all.arrays$post.means)
post.weights[which.max(post.weights)]

r=5
#r=which(colnames(b.test)=="Esophagus_Muscularis")
#r=which(colnames(b.test)=="Esophagus_Mucosa")
mu.gp1kl[r]
diag(U.gp1kl)[r]
(lower=pnorm(0,mean = mu.gp1kl[r],sd = sqrt(diag(U.gp1kl)[r]),lower.tail = T))
all.arrays$post.downs[k,r]
(upper=pnorm(0,mean = mu.gp1kl[r],sd = sqrt(diag(U.gp1kl)[r]),lower.tail = F))
all.arrays$post.ups[k,r]
1-max(lower,upper)


##show that this is lcose to lfsr because majority componenet
lfsr[j,r]


dim(post.weights)
dim(all.arrays$post.downs)
dim(all.arrays$post.ups)
total.down=post.weights%*%all.arrays$post.downs
total.up=post.weights%*%all.arrays$post.ups
plot(as.matrix(lfsr[j,]),as.matrix(apply(rbind(rbind(total.down,total.up)),2,function(x){1-max(x)})))

```

And here, we example singletons and distribution of number of strains signficant:
```{r}

singletons=which(rowSums(lfsr<0.05)==1)
barplot(apply(lfsr[singletons,],2,function(x){sum(x<0.05)}),names=colnames(t),las=2,main="Number of Singletons")
hist(apply(lfsr,1,function(x){sum(x<0.05)}),main="Number Strains Sig",breaks=seq(-0.5,9.5,by=1),las=3,xlab="Numeber of Strain < Thresh")

```


